### DISCUSSION TITLE
idea: option to reorder elements in single block

### DISCUSSION THREAD

**Discussion by psyafter** [voidlabs/mosaico#627] [Ideas]
There is any option to reorder elements in single template block?
For example I have block with 3 fields (f1, f2, f3).
I can add it to template, than I want reorder items in this block (when editing template)
For example I want following order (edited via style toolbox) - f2, f3, f1.

Solution can be:
1) add arrows "drag element" like with block ordering
OR
2) in style toolbox - add dropdown with order to each supported element (defined with unique property on element itself or any other way). changing option in dropdown will reorder it in block

**Comment by bago** [comment]
At this time there is no "element" concept. The template defines _blocks_, and each block define its own html using properties. There is no definition of elements at all.

In order to support this we would need to define element-containers and elements (like we do with block-container and blocks)

This idea would help the "social icons block" because we could let the user to sort the icons: the social icon block could be implemented with something more specific (like "collections support") because the html for each social icon is identical, so "elements have the same structure")

Can you detail your specific use case so I better understand it? 

Idea #68 is related to this

Major difference between #68 and this one is that #68 is a "two-level" where the "elements" would be "global" and could be added to any block, while this idea is to have each block zone (element-container) to define it's own elements and you just work inside that zone (add elements/hide elements/sort elements).

**Comment by psyafter** [comment]
Example with "social icons" is good. 
Container: ul, elements - li of ul.
So order should be order of li (each li can have different html).

In my case I want give a user a form with predefined html fields (name, age, email, phone) wrapped with labels or li,  but user should be able to reorder fields (li with content) like he prefer.

**Comment by bago** [comment]
Well, the social icons and the html fields requires 2 different implementations.

Speaking of data model , the social icons can be defined with an array of objects (social icons) where the objects are all of the same type. 
In order to support the second scenario, instead, one would need to store an array of objects and their "type" because their content may differ.

Also, for the social icons you also need an option to choose a new social network to be added or to remove one from the list (so add and remove in addition to sort).. I don't know if this is the same for the "form" solution.

**Comment by psyafter** [comment]
In "form" solution I do not need add new fields. Just reorder and show/hide (already implemented).

Why you need array of objects? 
It can be:
1) container description = css class or even mosaico attribute/property for referencing container (`<ul class="container1">`)
2) element = each element on first level of container (like in jquery sortable)
3) you can save initial order (count all elements on first level, so you have [0,1,2,3,...,COUNT-1]
than when order changing - just save new order [3,1,2,0,...] so you can restore the order of elements when restoring template from data

so it must have static html structure: 
`<container>
 <element>...</element>
 <element>...</element>
 <element>...</element>
</container>
`
and manipulation should be simple (ordering elements with their content)

pretty simple, isn't?

**Comment by psyafter** [comment]
any news about this feature?

**Comment by bago** [comment]
Thisi is not currently in our roadmap.
